File: app/src/main/java/com/solutions/alphil/zambiajobalerts/ui/savedjobs/SavedJobsViewModel.java
package com.solutions.alphil.zambiajobalerts.ui.savedjobs;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.solutions.alphil.zambiajobalerts.classes.Job;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class SavedJobsViewModel extends AndroidViewModel {

    private static final String PREFS_NAME = "saved_jobs_prefs";
    private static final String KEY_SAVED_JOBS = "saved_jobs_list";
    private final SharedPreferences prefs;
    private final Gson gson;

    private final MutableLiveData<List<Job>> _savedJobs = new MutableLiveData<>(new ArrayList<>());
    public LiveData<List<Job>> getSavedJobs() {
        return _savedJobs;
    }

    public SavedJobsViewModel(@NonNull Application application) {
        super(application);
        prefs = application.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        gson = new Gson();
        loadSavedJobs();
    }

    public void loadSavedJobs() {
        String json = prefs.getString(KEY_SAVED_JOBS, null);
        if (json != null) {
            Type type = new TypeToken<ArrayList<Job>>() {}.getType();
            List<Job> jobs = gson.fromJson(json, type);
            _savedJobs.setValue(jobs);
        } else {
            _savedJobs.setValue(new ArrayList<>());
        }
    }

    public void saveJob(Job job) {
        List<Job> currentJobs = _savedJobs.getValue();
        if (currentJobs == null) currentJobs = new ArrayList<>();

        // Check if already saved
        boolean alreadySaved = false;
        for (Job savedJob : currentJobs) {
            if (savedJob.getId() == job.getId()) {
                alreadySaved = true;
                break;
            }
        }

        if (!alreadySaved) {
            currentJobs.add(job);
            saveToPrefs(currentJobs);
            _savedJobs.setValue(currentJobs);
        }
    }

    public void removeJob(int jobId) {
        List<Job> currentJobs = _savedJobs.getValue();
        if (currentJobs != null) {
            currentJobs.removeIf(job -> job.getId() == jobId);
            saveToPrefs(currentJobs);
            _savedJobs.setValue(currentJobs);
        }
    }

    private void saveToPrefs(List<Job> jobs) {
        String json = gson.toJson(jobs);
        prefs.edit().putString(KEY_SAVED_JOBS, json).apply();
    }

    public boolean isJobSaved(int jobId) {
        List<Job> currentJobs = _savedJobs.getValue();
        if (currentJobs != null) {
            for (Job job : currentJobs) {
                if (job.getId() == jobId) return true;
            }
        }
        return false;
    }
}

File: app/src/main/java/com/solutions/alphil/zambiajobalerts/ui/savedjobs/SavedJobsViewModel.kt
package com.solutions.alphil.zambiajobalerts.ui.savedjobs

import android.app.Application
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.solutions.alphil.zambiajobalerts.classes.Job

class SavedJobsViewModel(application: Application) : AndroidViewModel(application) {
    private val prefs = application.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val gson = Gson()
    private val savedJobsLiveData = MutableLiveData<List<Job>>(emptyList())

    init {
        loadSavedJobs()
    }

    fun getSavedJobs(): LiveData<List<Job>> = savedJobsLiveData

    fun loadSavedJobs() {
        val json = prefs.getString(KEY_SAVED_JOBS, null)
        if (json != null) {
            val type = object : TypeToken<ArrayList<Job>>() {}.type
            savedJobsLiveData.value = gson.fromJson<List<Job>>(json, type) ?: emptyList()
        } else {
            savedJobsLiveData.value = emptyList()
        }
    }

    fun saveJob(job: Job) {
        val currentJobs = savedJobsLiveData.value.orEmpty().toMutableList()
        val alreadySaved = currentJobs.any { it.getId() == job.getId() }

        if (!alreadySaved) {
            currentJobs.add(job)
            saveToPrefs(currentJobs)
            savedJobsLiveData.value = currentJobs
        }
    }

    fun removeJob(jobId: Int) {
        val currentJobs = savedJobsLiveData.value?.toMutableList() ?: return
        currentJobs.removeAll { it.getId() == jobId }
        saveToPrefs(currentJobs)
        savedJobsLiveData.value = currentJobs
    }

    fun toggleJob(job: Job): Boolean {
        return if (isJobSaved(job.getId())) {
            removeJob(job.getId())
            false
        } else {
            saveJob(job)
            true
        }
    }

    private fun saveToPrefs(jobs: List<Job>) {
        prefs.edit().putString(KEY_SAVED_JOBS, gson.toJson(jobs)).apply()
    }

    fun isJobSaved(jobId: Int): Boolean =
        savedJobsLiveData.value.orEmpty().any { it.getId() == jobId }

    companion object {
        private const val PREFS_NAME = "saved_jobs_prefs"
        private const val KEY_SAVED_JOBS = "saved_jobs_list"
    }
}
